File: instrumentation/fragment/src/main/java/io/opentelemetry/android/instrumentation/fragment/RumFragmentLifecycleCallbacks.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.fragment;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import io.opentelemetry.android.instrumentation.common.ActiveSpan;
import io.opentelemetry.android.instrumentation.common.ScreenNameExtractor;
import io.opentelemetry.api.trace.Tracer;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public class RumFragmentLifecycleCallbacks extends FragmentManager.FragmentLifecycleCallbacks {
    private final Map<String, FragmentTracer> tracersByFragmentClassName = new HashMap<>();

    private final Tracer tracer;
    private final Supplier<String> lastVisibleScreen;
    private final ScreenNameExtractor screenNameExtractor;

    public RumFragmentLifecycleCallbacks(
            Tracer tracer,
            Supplier<String> lastVisibleScreen,
            ScreenNameExtractor screenNameExtractor) {
        this.tracer = tracer;
        this.lastVisibleScreen = lastVisibleScreen;
        this.screenNameExtractor = screenNameExtractor;
    }

    @Override
    public void onFragmentPreAttached(
            @NonNull FragmentManager fm, @NonNull Fragment f, @NonNull Context context) {
        super.onFragmentPreAttached(fm, f, context);
        getTracer(f).startFragmentCreation().addEvent("fragmentPreAttached");
    }

    @Override
    public void onFragmentAttached(
            @NonNull FragmentManager fm, @NonNull Fragment f, @NonNull Context context) {
        super.onFragmentAttached(fm, f, context);
        addEvent(f, "fragmentAttached");
    }

    @Override
    public void onFragmentPreCreated(
            @NonNull FragmentManager fm, @NonNull Fragment f, @Nullable Bundle savedInstanceState) {
        super.onFragmentPreCreated(fm, f, savedInstanceState);
        addEvent(f, "fragmentPreCreated");
    }

    @Override
    public void onFragmentCreated(
            @NonNull FragmentManager fm, @NonNull Fragment f, @Nullable Bundle savedInstanceState) {
        super.onFragmentCreated(fm, f, savedInstanceState);
        addEvent(f, "fragmentCreated");
    }

    @Override
    public void onFragmentViewCreated(
            @NonNull FragmentManager fm,
            @NonNull Fragment f,
            @NonNull View v,
            @Nullable Bundle savedInstanceState) {
        super.onFragmentViewCreated(fm, f, v, savedInstanceState);
        getTracer(f).startSpanIfNoneInProgress("Restored").addEvent("fragmentViewCreated");
    }

    @Override
    public void onFragmentStarted(@NonNull FragmentManager fm, @NonNull Fragment f) {
        super.onFragmentStarted(fm, f);
        addEvent(f, "fragmentStarted");
    }

    @Override
    public void onFragmentResumed(@NonNull FragmentManager fm, @NonNull Fragment f) {
        super.onFragmentResumed(fm, f);
        getTracer(f)
                .startSpanIfNoneInProgress("Resumed")
                .addEvent("fragmentResumed")
                .addPreviousScreenAttribute()
                .endActiveSpan();
    }

    @Override
    public void onFragmentPaused(@NonNull FragmentManager fm, @NonNull Fragment f) {
        super.onFragmentPaused(fm, f);
        getTracer(f).startSpanIfNoneInProgress("Paused").addEvent("fragmentPaused").endActiveSpan();
    }

    @Override
    public void onFragmentStopped(@NonNull FragmentManager fm, @NonNull Fragment f) {
        super.onFragmentStopped(fm, f);
        getTracer(f)
                .startSpanIfNoneInProgress("Stopped")
                .addEvent("fragmentStopped")
                .endActiveSpan();
    }

    @Override
    public void onFragmentSaveInstanceState(
            @NonNull FragmentManager fm, @NonNull Fragment f, @NonNull Bundle outState) {
        super.onFragmentSaveInstanceState(fm, f, outState);
    }

    @Override
    public void onFragmentViewDestroyed(@NonNull FragmentManager fm, @NonNull Fragment f) {
        super.onFragmentViewDestroyed(fm, f);
        getTracer(f)
                .startSpanIfNoneInProgress("ViewDestroyed")
                .addEvent("fragmentViewDestroyed")
                .endActiveSpan();
    }

    @Override
    public void onFragmentDestroyed(@NonNull FragmentManager fm, @NonNull Fragment f) {
        super.onFragmentDestroyed(fm, f);
        // note: this might not get called if the dev has checked "retainInstance" on the fragment
        getTracer(f).startSpanIfNoneInProgress("Destroyed").addEvent("fragmentDestroyed");
    }

    @Override
    public void onFragmentDetached(@NonNull FragmentManager fm, @NonNull Fragment f) {
        super.onFragmentDetached(fm, f);
        // this is a terminal operation, but might also be the only thing we see on app getting
        // killed, so
        getTracer(f)
                .startSpanIfNoneInProgress("Detached")
                .addEvent("fragmentDetached")
                .endActiveSpan();
    }

    private void addEvent(@NonNull Fragment fragment, String eventName) {
        FragmentTracer fragmentTracer =
                tracersByFragmentClassName.get(fragment.getClass().getName());
        if (fragmentTracer != null) {
            fragmentTracer.addEvent(eventName);
        }
    }

    private FragmentTracer getTracer(Fragment fragment) {
        FragmentTracer activityTracer =
                tracersByFragmentClassName.get(fragment.getClass().getName());
        if (activityTracer == null) {
            activityTracer =
                    FragmentTracer.builder(fragment)
                            .setTracer(tracer)
                            .setScreenName(screenNameExtractor.extract(fragment))
                            .setActiveSpan(new ActiveSpan(lastVisibleScreen))
                            .build();
            tracersByFragmentClassName.put(fragment.getClass().getName(), activityTracer);
        }
        return activityTracer;
    }
}

File: instrumentation/fragment/src/main/java/io/opentelemetry/android/instrumentation/fragment/RumFragmentLifecycleCallbacks.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.fragment

import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import io.opentelemetry.android.instrumentation.common.ActiveSpan
import io.opentelemetry.android.instrumentation.common.ScreenNameExtractor
import io.opentelemetry.api.trace.Tracer

internal class RumFragmentLifecycleCallbacks(
    private val tracer: Tracer,
    private val lastVisibleScreen: () -> String?,
    private val screenNameExtractor: ScreenNameExtractor,
) : FragmentManager.FragmentLifecycleCallbacks() {
    private val tracersByFragmentClassName: MutableMap<String, FragmentTracer> = mutableMapOf()

    override fun onFragmentPreAttached(
        fm: FragmentManager,
        f: Fragment,
        context: Context,
    ) {
        super.onFragmentPreAttached(fm, f, context)
        getTracer(f).startFragmentCreation().addEvent("fragmentPreAttached")
    }

    override fun onFragmentAttached(
        fm: FragmentManager,
        f: Fragment,
        context: Context,
    ) {
        super.onFragmentAttached(fm, f, context)
        addEvent(f, "fragmentAttached")
    }

    override fun onFragmentPreCreated(
        fm: FragmentManager,
        f: Fragment,
        savedInstanceState: Bundle?,
    ) {
        super.onFragmentPreCreated(fm, f, savedInstanceState)
        addEvent(f, "fragmentPreCreated")
    }

    override fun onFragmentCreated(
        fm: FragmentManager,
        f: Fragment,
        savedInstanceState: Bundle?,
    ) {
        super.onFragmentCreated(fm, f, savedInstanceState)
        addEvent(f, "fragmentCreated")
    }

    override fun onFragmentViewCreated(
        fm: FragmentManager,
        f: Fragment,
        v: View,
        savedInstanceState: Bundle?,
    ) {
        super.onFragmentViewCreated(fm, f, v, savedInstanceState)
        getTracer(f).startSpanIfNoneInProgress("Restored").addEvent("fragmentViewCreated")
    }

    override fun onFragmentStarted(
        fm: FragmentManager,
        f: Fragment,
    ) {
        super.onFragmentStarted(fm, f)
        addEvent(f, "fragmentStarted")
    }

    override fun onFragmentResumed(
        fm: FragmentManager,
        f: Fragment,
    ) {
        super.onFragmentResumed(fm, f)
        getTracer(f)
            .startSpanIfNoneInProgress("Resumed")
            .addEvent("fragmentResumed")
            .addPreviousScreenAttribute()
            .endActiveSpan()
    }

    override fun onFragmentPaused(
        fm: FragmentManager,
        f: Fragment,
    ) {
        super.onFragmentPaused(fm, f)
        getTracer(f).startSpanIfNoneInProgress("Paused").addEvent("fragmentPaused").endActiveSpan()
    }

    override fun onFragmentStopped(
        fm: FragmentManager,
        f: Fragment,
    ) {
        super.onFragmentStopped(fm, f)
        getTracer(f)
            .startSpanIfNoneInProgress("Stopped")
            .addEvent("fragmentStopped")
            .endActiveSpan()
    }

    override fun onFragmentViewDestroyed(
        fm: FragmentManager,
        f: Fragment,
    ) {
        super.onFragmentViewDestroyed(fm, f)
        getTracer(f)
            .startSpanIfNoneInProgress("ViewDestroyed")
            .addEvent("fragmentViewDestroyed")
            .endActiveSpan()
    }

    override fun onFragmentDestroyed(
        fm: FragmentManager,
        f: Fragment,
    ) {
        super.onFragmentDestroyed(fm, f)
        // note: this might not get called if the dev has checked "retainInstance" on the fragment
        getTracer(f).startSpanIfNoneInProgress("Destroyed").addEvent("fragmentDestroyed")
    }

    override fun onFragmentDetached(
        fm: FragmentManager,
        f: Fragment,
    ) {
        super.onFragmentDetached(fm, f)
        // this is a terminal operation, but might also be the only thing we see on app getting
        // killed, so
        getTracer(f)
            .startSpanIfNoneInProgress("Detached")
            .addEvent("fragmentDetached")
            .endActiveSpan()
    }

    private fun addEvent(
        fragment: Fragment,
        eventName: String,
    ) {
        val fragmentTracer = tracersByFragmentClassName[fragment.javaClass.name]
        fragmentTracer?.addEvent(eventName)
    }

    private fun getTracer(fragment: Fragment): FragmentTracer =
        tracersByFragmentClassName.getOrPut(fragment.javaClass.name) {
            FragmentTracer(
                fragment,
                tracer,
                screenNameExtractor.extract(fragment),
                ActiveSpan(lastVisibleScreen),
            )
        }
}
